File: leetcode/120.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
private:
  unordered_map<string, int> cache;
  int dfs(vector<vector<int>> &triangle, int d, int p) {
    if (d >= triangle.size()) {
      return 0;
    }
    string key = to_string(d) + ',' + to_string(p);
    if (cache.count(key) == 0) {
      cache[key] = triangle[d][p] +
                   min(dfs(triangle, d + 1, p), dfs(triangle, d + 1, p + 1));
    }
    return cache[key];
  }

public:
  int minimumTotal(vector<vector<int>> &triangle) {
    return dfs(triangle, 0, 0);
  }

  // 自顶向上
  int minimumTotal_(vector<vector<int>> &triangle) {
    for (int i = triangle.size() - 1; i > 0; i--) {
      for (int j = 0; j < triangle[i].size() - 1; j++) {
        triangle[i - 1][j] += min(triangle[i][j], triangle[i][j + 1]);
      }
    }
    return triangle[0][0];
  }
};

int main() {
  vector<vector<int>> triangle = {{2}, {3, 4}, {6, 5, 7}, {4, 1, 8, 3}};
  cout << Solution().minimumTotal(triangle) << endl;
  return 0;
}

File: leetcode/120.py
from typing import List


class Solution:

    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''
        暴力解法
        '''
        return self.dfs(triangle, len(triangle), 0, 0)

    def dfs(self, triangle: List[List[int]], size: int, row: int,
            col: int) -> int:
        if size == row:
            return 0
        return min(self.dfs(triangle, size, row + 1, col),
                   self.dfs(triangle, size, row + 1,
                            col + 1)) + triangle[row][col]


class Solution1:

    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''
        缓存备忘录
        '''
        self.help = {}
        return self.dfs(triangle, len(triangle), 0, 0)

    def dfs(self, triangle: List[List[int]], size: int, row: int,
            col: int) -> int:
        if size == row:
            return 0
        key = (
            row,
            col,
        )
        if key in self.help:
            return self.help[key]
        self.help[key] = min(self.dfs(triangle, size, row + 1, col),
                             self.dfs(triangle, size, row + 1,
                                      col + 1)) + triangle[row][col]
        return self.help[key]


class Solution2:

    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''
        动态规划
        '''
        for i in range(len(triangle) - 2, -1, -1):
            for j in range(len(triangle[i])):
                # triangle[i][j] += min(triangle[i + 1][j],
                #                       triangle[i + 1][j + 1])
                triangle[i][j] += min(triangle[i + 1][j:j + 2])
        return triangle[0][0]


if __name__ == "__main__":
    for i in range(5, -1, -1):
        print(i)